test/powershell/engine/Api/MultithreadedEngine.Tests.ps1

```diff
@@ -0,0 +1,198 @@
+                        $Exception.ErrorRecord.FullyQualifiedErrorId
+                        break
+                    }
+                }
+            }
+        }
+
+        It 'can invoke a single script asynchronously, but only in a new runspace' {
+            $ps = [powershell]::Create()
+            $r = InvokeAsyncHelper -PowerShell $ps -Wait
+            $r.Status | Should -Be ([System.Threading.Tasks.TaskStatus]::RanToCompletion)
+            $r.IsCompletedSuccessfully | Should -Be $true
+        }
+
+        It 'handles terminating errors properly when invoked asynchronously in a new runspace' {
+            $ps = [powershell]::Create()
+            $sb = {
+                $r = $ps.AddScript(@'
+try {
+    Get-Process -InvalidParameter 42 -ErrorAction Stop
+} catch {
+    throw
+}
+'@).InvokeAsync()
+                Set-Variable -Name r -Scope 2 -Value $r
+                [System.Threading.Tasks.Task]::WaitAll(@($r))
+                $r
+            }
+            # This test is designed to gracefully fail with an error when invoked asynchronously.
+            $err = { $sb.Invoke() } | Should -Throw -ErrorId 'AggregateException' -PassThru
+            $r.IsFaulted | Should -Be $true
+            $err.Exception.InnerException.InnerException.ErrorRecord.FullyQualifiedErrorId | Should -BeExactly 'NamedParameterNotFound,Microsoft.PowerShell.Commands.GetProcessCommand'
+        }
+
+        It 'cannot invoke a single script asynchronously in a runspace that has not been opened' {
+            $rs = [runspacefactory]::CreateRunspace()
+            $ps = [powershell]::Create($rs)
+            $r = $ps.AddScript('@(1..10).foreach{Start-Sleep -Milliseconds 500}').InvokeAsync()
+            # This test is designed to fail. You cannot invoke PowerShell asynchronously
+            # in a runspace that has not been opened.
+            $r.IsFaulted | Should -Be $true
+            $r.Exception.InnerException -is [System.Management.Automation.Runspaces.InvalidRunspaceStateException] | Should -Be $true
+            $r.Exception.InnerException.CurrentState | Should -Be 'BeforeOpen'
+            $r.Exception.InnerException.ExpectedState | Should -Be 'Opened'
+        }
+
+        It 'cannot invoke a single script asynchronously in a runspace that is busy' {
+            $rs = [runspacefactory]::CreateRunspace()
+            $rs.Open()
+            $psBusy = [powershell]::Create($rs)
+            $r = $psBusy.AddScript('@(1..5).foreach{Start-Sleep -Milliseconds 500}').InvokeAsync()
+            $rs.RunspaceAvailability | Should -Be 'Busy'
+            $ps = [powershell]::Create($rs)
+            $sb = {
+                InvokeAsyncHelper -PowerShell $ps -Wait
+            }
+            # This test is designed to fail. You cannot invoke PowerShell asynchronously
+            # in a runspace that is busy, because pipelines cannot be run concurrently.
+            $err = { $sb.Invoke() } | Should -Throw -ErrorId 'AggregateException' -PassThru
```

Comment:
This may not be reliable too, as the previous asynchronous execution might already finish ...

Reply:
Based on other update, we should be good.